zio-http/shared/src/main/scala/zio/http/endpoint/openapi/JsonSchema.scala

```diff
@@ -1248,23 +1285,79 @@ object JsonSchema {
     def required(required: Chunk[java.lang.String]): Object =
       this.copy(required = required)
 
+    private def reconcileTwoOptions[T](left: Option[T], right: Option[T], noneIfOne: Boolean)(
+      combine: (T, T) => Option[T],
+    ): Option[T] = {
+      if (noneIfOne) left.fold(Option.empty[T])(l => right.fold(Option.empty[T])(r => combine(l, r)))
+      else left.fold(right)(l => right.fold(left)(r => combine(l, r)))
+    }
+
+    private def someWhenEq[T](l: T, r: T): Option[T] =
+      if (l == r) Some(l) else None
+
+    private def combinePatterns(lPattern: Pattern, rPattern: Pattern): Some[Pattern] =
+      if (lPattern == rPattern) Some(lPattern)
+      else {
+        // validate both patterns match using lookahead assertions
+        Some(Pattern("(?=" + lPattern + ")(?=" + rPattern + ")"))
+      }
+
+    private def wrap[T](f: (T, T) => T): (T, T) => Option[T] = (l, r) => Some(f(l, r))
+
     private def combineAdditionalProperties(
       left: Either[Boolean, JsonSchema],
       right: Either[Boolean, JsonSchema],
     ): Either[Boolean, JsonSchema] =
       (left, right) match {
-        case (Left(false), _)            => Left(false)
-        case (_, Left(_))                => left
-        case (Left(true), _)             => right
-        case (Right(left), Right(right)) =>
-          Right(AllOfSchema(Chunk(left, right)))
+        case (Left(false), _)                 => Left(false)
+        case (_, Left(_))                     => left
+        case (Left(true), _)                  => right
+        case (Right(lSchema), Right(rSchema)) =>
+          val (leftKey, lAnnotations)  = Object.extractKeySchemaFromAnnotations(lSchema)
```

Comment:
I am not sure we should make it this complicated. Maybe it is okay we say we only support a certain sub set. Like only one key annotated schema?

Reply:
only one key means 2 dictionaries: 1 unconstrained, and 1 has a schema. in this case, code resorts to `None`, since we cannot have any schema.
The only applicable case is when we have 2 schemas.
I chose to implement "intersection" semantics when possible, over "union".

This is an important decision, which I'm not sure about.
What's your take? should we relax the key schema to be a union of both, or restrict to be the intersection of both?

Also, when does combine is called anyway?

Reply:
I think, id we have something that we can't resolve successfully, we should throw. At least something like not impl exception. Dropping it silently is hell for debugging

Reply:
In general I agree with the principle, not sure it applies here.
- We're not dropping silently
- it's an extension that is ignored by OpenAPI (anything that starts with `x-` is passed on transparently), thus I think we should avoid throwing as much as possible.

lets start with the obvious: if both are the same (either None, or identical Some(x)) - preserve.

The question is when they differ.
If only 1 is defined, i.e. combining 2 dictionaries, one restricted (e.g. UUID keys), the other is not.
So I assume the combined should not be constrained?

If both are defined, then some properties can be reconciled (e.g. min/max length or pattern), while others less so (format).
So I tried to do the best here.

What I lack, is the context of when combining actually happens.

Reply:
@987Nabil IIUC, combine is called only when generating JsonSchema from zio Schema, i.e. when we define the endpoint in scala, and generate the schema from it.
If that's the case, I think that my decision to do an "intersection" was wrong, and I should do a "union".
I.e. max of maxLength (instead min of maxLength), `|` instead of lookahead assertions in pattern, etc'…

I'll change accordingly.